### PR TITLE
make the headers property editable

### PR DESCRIPTION
**Pull request by itisnajim**
## What kind of change does this PR introduce?

This PR introduces a feature that allows for greater customization of HTTP headers in the Supabase Dart library.

## What is the current behavior?

Currently, the library uses a private _headers field to store HTTP headers for every request. Users can't modify these headers outside of the library code.

## What is the new behavior?

This PR adds a getter and setter for the _headers field, allowing users to customize headers for specific requests, such as adding or removing headers as needed.

## Additional context

This change will give users more control over the headers used in their requests, making the Supabase Dart library more flexible and adaptable to their needs. No relevant issues were identified. No screenshots are necessary as this is a code change.

Here's an example of how the new feature can be used in the RLS with the device_uid() function:

Suppose you have a users table and want to allow read access to unauthenticated users only if they provide a valid x-device-uuid header. You can create a policy like this:

```sql
CREATE POLICY allow_read_unauthenticated ON users
  FOR SELECT
  USING (
    (auth.role() = 'anonymous' AND device_uid() IS NOT NULL) OR
    auth.uid() = id
  );
```

This policy allows unauthenticated users to read a row only if they provide a valid x-device-uuid header, or if the row belongs to them (identified by their uid).

this is the definition of device_uid() if someone need it 🙂 ✨
```sql
CREATE OR REPLACE FUNCTION device_uid() RETURNS uuid LANGUAGE sql STABLE AS $function$
SELECT NULLIF(
    current_setting('request.headers', true)::json->>'x-device-uuid',
    ''
  )::uuid;
$function$;
```

### CODE REVIEW

File: lib/src/supabase_client.dart [hunk 1]
```diff
@@ -42,6 +42,21 @@ class SupabaseClient {
...
+  /// Setter for the HTTP headers
+  set headers(Map<String, String> value) {
+    final keysToRemove =
+        _headers.keys.toSet().where((key) => !value.containsKey(key));
+
+    _headers
+      ..addAll(value)
+      ..removeWhere((key, _) => keysToRemove.contains(key));
+  }
+
```
Comment (dshukertjr) [inline lib/src/supabase_client.dart:59]: I wonder if this this is the same as making `_headers` variable public like `headers`.

**Review by dshukertjr — COMMENTED** [review]
Thanks for this PR, and your use case with device ID is very convincing.

I have [denied a use case where users can edit their own headers](https://github.com/supabase/supabase-flutter/issues/179), but I admit that it was probably not the wisest decisions, and think having control over the header does make sense.

File: lib/src/supabase_client.dart (hunk 1 above)
Comment (itisnajim) [reply lib/src/supabase_client.dart:59]: addAll in: 
```dart
_headers
      ..addAll(value)
      ...
```
will update existing values if the key already exists in the map. This means that if a key already exists, the corresponding value will be overwritten with the new value provided in the `value` argument map.

Therefore, the getter and setter mentioned in the previous code snippet are 100% equivalent to making `_headers` mutable (non final).

Comment (Vinzent03) [reply lib/src/supabase_client.dart]: Even with the `final` keyword, `_headers` is mutable. You just can't reassign it, but you can change the existing object.

**Review by Vinzent03 — CHANGES_REQUESTED** [review]
This won't work, because once the `storage`, `functions`, `realtime` are created,  changing the `headers` field won't affect the sub libraries. I think we need to do it like with the existing `setAuth()` in `_handleTokenChanged()`.

**Comment by itisnajim** [comment]
> This won't work, because once the `storage`, `functions`, `realtime` are created, changing the `headers` field won't affect the sub libraries. I think we need to do it like with the existing `setAuth()` in `_handleTokenChanged()`.

to do so we have to update storage-dart, functions and realtime packages first, by making _headers property in each package a non-final property adding a setter to each one (this needed to control or to add default headers if one is missing or so ! ..)

then updating SupabaseClient to the following:
```dart
Map<String, String> _headers;

set headers(Map<String, String> value) {
    final keysToRemove =
        _headers.keys.toSet().where((key) => !value.containsKey(key));

    _headers
      ..addAll(value)
      ..removeWhere((key, _) => keysToRemove.contains(key))
      // to make sure that the default headers are always included
      ..addAll(Constants.defaultHeaders); 

    // using the setters in each package to update the headers
    realtime.headers = headers;
    functions.headers = headers;
    storage.headers = headers;
  }
  ```

**Comment by dshukertjr** [comment]
@itisnajim 
I like the idea of always including the default header! Maybe we can add the default headers first so that the developers can override it if they want to. Something like this should do it I think.

```dart
set headers(Map<String, String> value) {
    _headers
      ..clear() // instead of finding keys to remove and removing them, we can probably just clear the entire thing!
      ..addAll(Constants.defaultHeaders)
      ..addAll(value);

    // using the setter in each package to update the headers
    realtime.headers = _headers;
    functions.headers = _headers;
    storage.headers = _headers;
  }
```

**Comment by itisnajim** [comment]
> @itisnajim I like the idea of always including the default header! Maybe we can add the default headers first so that the developers can override it if they want to. Something like this should do it I think.
> 
> ```dart
> set headers(Map<String, String> value) {
>     _headers
>       ..clear() // instead of finding keys to remove and removing them, we can probably just clear the entire thing!
>       ..addAll(Constants.defaultHeaders)
>       ..addAll(value);
> 
>     // using the setter in each package to update the headers
>     realtime.headers = _headers;
>     functions.headers = _headers;
>     storage.headers = _headers;
>   }
> ```

I've opened PRs in each package
https://github.com/supabase/realtime-dart/pull/63
https://github.com/supabase/storage-dart/pull/57
https://github.com/supabase/functions-dart/pull/19

after the merge i'll update for this `supabase-dart` too.

File: lib/src/supabase_client.dart [hunk 2]
```diff
@@ -64,7 +69,7 @@ class SupabaseClient {
     this.supabaseKey, {
     String? schema,
     bool autoRefreshToken = true,
-    Map<String, String> headers = Constants.defaultHeaders,
```
Comment (itisnajim) [inline lib/src/supabase_client.dart:67]: this make the headers unmodifiable ! removed.

**Comment by Vinzent03** [comment]
The conclusion from the discussion in https://github.com/supabase/functions-dart/pull/19 was 

> To complete the implementation that we have, why don't we just expose a getter for headers in all of the sub libraries, and introduce a setter and a getter on supabase-dart. Within the method of setter on supabase-dart, we can update each sub libraries' headers.

`realtime-dart`, `postgrest-dart` and `storage-dart` already have a public `headers` field, so their pr is not necessary for this one. For `functions-dart` https://github.com/supabase/functions-dart/pull/19 exists, but I think for `gotrue-dart` there needs to be a fix as well.

**Comment by itisnajim** [comment]
> The conclusion from the discussion in [supabase/functions-dart#19](https://github.com/supabase/functions-dart/pull/19) was
> 
> > To complete the implementation that we have, why don't we just expose a getter for headers in all of the sub libraries, and introduce a setter and a getter on supabase-dart. Within the method of setter on supabase-dart, we can update each sub libraries' headers.
> 
> `realtime-dart`, `postgrest-dart` and `storage-dart` already have a public `headers` field, so their pr is not necessary for this one. For `functions-dart` [supabase/functions-dart#19](https://github.com/supabase/functions-dart/pull/19) exists, but I think for `gotrue-dart` there needs to be a fix as well.

I did a PR for gotrue-dart with the fix, which you can find here: https://github.com/supabase/gotrue-dart/pull/137.

After the merge of gotrue-dart, do you think any additional changes will be needed for supabase-dart/pull/185? if so i'm here  😊

**Comment by Vinzent03** [comment]
Do we continue here with adding a setter as well or do we just keep the getter as it currently is?

**Comment by itisnajim** [comment]
> Do we continue here with adding a setter as well or do we just keep the getter as it currently is?

if we add a setter we have to ensure all the sub packages to have a setter so we can use within the parent setter (supabase-dart setter, but functions-dart and gotrue-dart have only a getter) or work around this to use a trick like : `functionsClient.headers..clear()..addAll(supabase._headers)` inside supabase-dart setter, additionally, we need a way (code snippet) that can apply the new headers in the realtime-dart client. 

What do you think about these considerations?

**Comment by Vinzent03** [comment]
We already agreed on only providing a getter in the sub libraries. `headers` is already public in realtime-dart, so there are no changes necessary. The channel reconnecting should be done by the user.

**Review by Vinzent03 — APPROVED** [review]
I think this is ready now. @dshukertjr What do you think?

**Review by dshukertjr — APPROVED** [review]
Looks Good!

**Pull request merged**